libcnb/src/layer/struct_api/handling.rs

```diff
@@ -0,0 +1,520 @@
+use crate::layer::shared::{
+    delete_layer, read_layer, replace_layer_metadata, replace_layer_types, ReadLayerError,
+    WriteLayerError,
+};
+use crate::layer::{
+    EmptyLayerCause, InspectRestoredAction, IntoAction, InvalidMetadataAction, LayerError,
+    LayerRef, LayerState,
+};
+use crate::Buildpack;
+use libcnb_common::toml_file::read_toml_file;
+use libcnb_data::generic::GenericMetadata;
+use libcnb_data::layer::LayerName;
+use libcnb_data::layer_content_metadata::{LayerContentMetadata, LayerTypes};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::marker::PhantomData;
+use std::path::{Path, PathBuf};
+
+pub(crate) fn handle_layer<B, M, MA, RA, MAC, RAC>(
+    layer_types: LayerTypes,
+    invalid_metadata: &dyn Fn(&GenericMetadata) -> MA,
+    inspect_restored: &dyn Fn(&M, &Path) -> RA,
```

Comment:
These need renaming to `invalid_metadata_action` / `restored_layer_action` too (not using suggested changes feature since there are a few more places below that will need updating as a result).